src/ConfigHandler.cpp

```diff
@@ -10,9 +10,29 @@
 
 using namespace std;
 
-void ConfigHandler::loadConfig(string filename) {
-  cout << "loadConfig() in ConfigHandler" << endl;
+/*  Initialize static member */
+shared_ptr<ConfigHandler> ConfigHandler::mInstance{nullptr};
+
+ConfigHandler& ConfigHandler::GetInstance() {
+  if (mInstance == nullptr)
+    mInstance = shared_ptr<ConfigHandler>(new ConfigHandler);
+  return *mInstance;
```

Comment:
as requested a in depth explanation:
smart pointers are intended for unique/shared of a heap allocated object. 
they promise two things:
1. The object gets automatically freed as soon as all owners go out of scope
2. You are always guaranteed to access a valid object (i.e. not already freed or similar)   

BUT as soon as you access the underlying object directly you can cause all these troubles again. 

If you only copy the object nothing bad can happen. If you keep a reference (which can be the case with the `ConfigHandler`) none of the guarantees of smart pointers hold true for this reference.

Meaning that if you hold a reference while the last owner (smart pointer) gets destroyed, you will have an invalid reference -> UNDEFINED BEHAVIOR (= a litte bad)

Solution: Either go full `shared_ptr` and return a (const) reference to it, or go back to raw pointers in which case no guarantees are given anyway and the object is destroyed at program termination (without an explicit `delete`)

Reply:
another point which should be considered: since we will access the `ConfigHandler` from more than 1 threads concurrently we either should write access (= calls to non-const functions) or lock the functions. Since we do not intend to write something anyway you could return a `const std::shared_ptr<ConfigHandler>` (without reference! otherwise we would get in troubles with invalid references again)

Reply:
how would i resolve the issue?
`return shared_ptr<shared_ptr<ConfigHandler>>(mInstance)` ?

Reply:
`return mInstance`  and return type of the function is `std::shared_ptr<ConfigHandler>  getInstance()` you can also add a `const &` at the return type

Reply:
as @janeson332 said, `std::shared_ptr<ConfigHandler> const getInstance()` with `return mInstance;` should do best.
avoid a reference for the reasons above.